Document/0x06g-Testing-Network-Communication.md

```diff
@@ -2,7 +2,20 @@
 
 Almost every iOS app acts as a client to one or more remote services. As this network communication usually takes place over untrusted networks such as public Wi-Fi, classical network based-attacks become a potential issue.
 
-Most modern mobile apps use variants of HTTP based web-services, as these protocols are well-documented and supported. On iOS, the `NSURLConnection` class provides methods to load URL requests asynchronously and synchronously.
+Most modern mobile apps use variants of HTTP based web-services, as these protocols are well-documented and supported.
+Since iOS 12.0 the [Network framework](https://developer.apple.com/documentation/network "API Reference Network") and the [`URLSession`](https://developer.apple.com/documentation/foundation/urlsession "API Reference URLSession") class provide methods to load network and URL requests asynchronously and synchronously. Older iOS versions can utilize the [Sockets API](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/NetworkingTopics/Articles/UsingSocketsandSocketStreams.html "Using Sockets and Socket Streams").
+
+### Network Framework
+
+`Network framework` was introduced at [The Apple Worldwide Developers Conference (WWDC)](https://developer.apple.com/videos/play/wwdc2018/715 "Introducing Network.framework: A modern alternative to Sockets") in 2018 and is a replacement to the Sockets API. This low-level networking framework provides classes to send and receive data with built in dynamic networking, security and performance support.
+
+TLS 1.3 is enabled by default in Network framework, if the argument `using: .tls` is used. It is the preferred option over the legacy [Secure Transport](https://developer.apple.com/documentation/security/secure_transport "API Reference Secure Transport") framework.
+
+### URLSession
+
+`URLSession` was built upon the Network framework and utilizes the same transport services. The class also uses TLS 1.3 by default, if the endpoint is HTTPS. `URLSession` should be used instead of the Network framework for HTTP and URL connections.
```

Comment:
i think this is a very nice point where we can add value here, that's why I insist. Imagine I'm a developer and I don't really know what's better of terms of security. As Apple states: "Use this framework when you need direct access to protocols like TLS, TCP, and UDP for your custom application protocols. " I'd add "ONLY' to that statement. I see `URLSession` as the safer option for _most of the cases_: you just want to retrieve some data (weather, kitten memes... you name it). If an careless developer would use the Network framework instead, wouldn't it present some misconfiguration risks?

it would be interesting to understand those risks, that way we have a clear statement from the security point of view about why we recommend to use `URLSession` for any URL-based resources. What do you think? :)

Reply:
Thanks @cpholguera, absolutely agree!  

As you can see in the Network framework code sample, there is a lot of boiler plate to be done by the developer just to send and receive HTTP-based requests in `Network` over `URLSession`. Additionally, URLSession has an immediate security benefit by not requiring the developer to specify `.tls` or `.tcp` and instead utilizing the protocol/scheme of the URL. 

Do you want me to try and condense the above and add in? 

I can also do more research (unfortunately there is not a lot of documentation/resources on Network framework), but I think the inherent risks are due to the added complexity of establishing and communicating over a Network framework connection.

Reply:
yes please, a condensed conclusion would help a lot here. The security relevant information already digested for them 👌 thanks!

Reply:
@cpholguera  and @sushi2k , what do l think about something like this for clarifying when to use Network framework or URLSession? 

... 


### URLSession

`URLSession` was built upon the Network framework and utilizes the same transport services. The class also uses TLS 1.3 by default, if the endpoint is HTTPS.

`URLSession` should be used for HTTP and HTTPS connections, instead of utilizing Network framework directly. The class natively supports both URL schemes and is optimized for such connections. It and requires less boilerplate code, reducing the propensity for errors and ensures default secure connections by inspecting the scheme of a URL, instead of requiring the connection protocol to be specified. Network framework should only be used when there are low-level and/or advanced networking requirements.

Reply:
Sounds good to me! Please keep referring to is as "The Network framework" instead of only "Network framework" to be consistent. Thanks for taking the time to think about it ;)

Reply:
Thank you @cpholguera and @sushi2k for all your help and support!  I have pushed those changes for your review.

Reply:
Reads really good! Thanks @cak